packages/core/src/indexing/service.ts

```diff
@@ -1,1187 +1,724 @@
-            }, block=${data.checkpoint.blockNumber}, eventIndex=${
-              data.checkpoint.eventIndex
-            } error=${`${error.name}: ${error.message}`})`,
-          });
-          await this.indexingStore.revert({
-            checkpoint: data.checkpoint,
-          });
-        }
-      }
+    // set total seconds
+    if (indexingService.lastEventCheckpoint !== undefined) {
+      indexingService.common.metrics.ponder_indexing_total_seconds.set(
+        indexingService.lastEventCheckpoint.blockTimestamp -
+          indexingService.firstEventCheckpoint.blockTimestamp,
+      );
     }
+  }
 
-    await this.loadingMutex.runExclusive(async () => {
-      const loadKeys = this.getLoadKeys();
+  for (let i = 0; i < events.length; i++) {
+    const event = events[i];
+    if (indexingService.isKilled) return { status: "killed" };
+    indexingService.eventCount++;
 
-      await Promise.all(
-        loadKeys.map((key) => this.loadIndexingFunctionTasks(key)),
-      );
+    indexingService.common.logger.trace({
+      service: "indexing",
+      msg: `Started indexing function (event="${event.contractName}:${event.eventName}", checkpoint=${event.encodedCheckpoint})`,
     });
 
-    this.enqueueLogEventTasks();
-  };
-
-  private createEventQueue = () => {
-    const indexingFunctionWorker: Worker<IndexingFunctionTask> = async ({
-      task,
-    }) => {
-      // This is a hack to ensure that the eventsProcessed method is called and updates
-      // the UI when using SQLite. It also allows the process to GC and handle SIGINT events.
-      // It does, however, slow down event processing a bit. Too frequent waits cause massive performance loses.
-      if (Math.floor(Math.random() * 100) === 69) await wait(0);
-
-      switch (task.kind) {
-        case "SETUP": {
-          await this.executeSetupTask(task);
-          break;
-        }
-        case "LOG": {
-          await this.executeLogEventTask(task);
-          break;
+    switch (event.type) {
+      case "log": {
+        const result = await executeLog(indexingService, { event });
+        if (result.status === "error") {
+          return { status: "error", error: result.error };
         }
+        break;
       }
-    };
-
-    this.queue = createQueue({
-      worker: indexingFunctionWorker,
-      options: {
-        concurrency: 10,
-        autoStart: false,
-      },
-    });
-  };
 
-  /**
-   * Load a batch of indexing function tasks from the sync store into memory.
-   */
-  loadIndexingFunctionTasks = async (key: string) => {
-    const state = this.indexingFunctionStates[key];
-    const tasks = state.loadedTasks;
-
-    // TODO: Deep copy these.
-    const fromCheckpoint = state.tasksLoadedToCheckpoint;
-    const toCheckpoint = this.syncService.checkpoint;
-
-    if (
-      isCheckpointGreaterThanOrEqualTo(fromCheckpoint, toCheckpoint) &&
-      state.lastEventCheckpoint !== undefined
-    ) {
-      return;
+      // default:
+      //   neva(event);
```

Comment:
?

Reply:
This code is needed if you have more than one type of event to index

Reply:
Got it, makes sense!